### PR TITLE
[MRG] Add functionality to record current from all cells

### LINKED ISSUE #187: Make cell currents a model output
HNN GUI saves cell currents for 'L2Pyr_soma' and  'L5Pyr_soma' to the file `i.txt`, but doesn't currently plot them. Adding the functionality to hnn-core to calculate these currents would allow us to explore ways to represent/plot this data in the future.

Function to record currents (called for L2Pyr and L5Pyr):
https://github.com/jonescompneurolab/hnn/blob/bd2abe4007160a5186856bc17dd87ae62055b383/cell.py#L180-L194

Needs to be aggregated for each type of cell: https://github.com/jonescompneurolab/hnn/blob/bd2abe4007160a5186856bc17dd87ae62055b383/network.py#L370-L387

The above function is called after a simulation completes:
https://github.com/jonescompneurolab/hnn/blob/d4c4cf94aa1db8517b99b0e344aa852c16125f50/run.py#L435

After this functionality exists to retrieve currents as a model output (same mechanism as somatic voltages?), the code to save the output to `i.txt` can be added back to HNN GUI. It would be a cleaner version of this code:
https://github.com/jonescompneurolab/hnn/blob/d4c4cf94aa1db8517b99b0e344aa852c16125f50/run.py#L140-L145

### PR DESCRIPTION
**Pull request by ntolley**
closes #187. Starting this PR early to keep track of modifications discussed in #193 

**To-do** 
- [x] Merge `record_current()` and `record_voltage()` into single function
- [x] Merge `aggregate_currents()` and `aggregate_voltages()` into single function
- [x] ~~Modify dipole calculation to utilize individual cell current recordings?~~
- [x] Add tests to check current/voltage waveform recordings agree with `spiketimes`
- [x] Rename `Spikes` class to `CellResponse`

### CODE REVIEW

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:66]: is there such a parameter in the `params` file? If not, let's not invent it ...

Comment (ntolley) [reply hnn_core/dipole.py:66]: So this was posed as the workaround for enabling/disabling `vsoma` recordings. Do you think this should be an attribute of the `Network` instead?

Comment (rythorpe) [reply hnn_core/dipole.py:66]: My vote is to leave `record_isoma` as an argument of `simulate_dipole()` without creating a new param. Same with `record_vsoma`, however, since it's already an argument in the default param file as well as the GUI (right?), I guess it's fine to leave `record_vsoma` in there until after GUI integration. I'm on an anti-params-file kick at the moment so that we can hopefully refocus and simplify the use-case of parameters stored in param files :)

Comment (ntolley) [reply hnn_core/dipole.py:66]: Ah I missed this. `save_vsoma` is contained inside `default.json` but I think that's just leftover from copying the GUI params directly. 

Currently this is just an awkward way to pass `record_vsoma` to the `network_builder.py`. The params variable is just a convenient way to pass these variables, but neither is specified in the actual `.params` file. 

I'm not a fan of adding extra arguments to `_BACKEND.simulate(net)` so the alternative would be adding an attribute to `Network` which can be accessed by `network_builder.py` later on.

Comment (jasmainak) [reply hnn_core/dipole.py:66]: we have a bit of a mess with parallel backends. Maybe you get away with `net.params` now. But we'll have to bite the bullet at some point and clean it up. If we don't touch the code, we won't realize how to improve it :)

Comment (rythorpe) [reply hnn_core/dipole.py:66]: Oh I see. Yeah that will be a bit of a mess, but I feel like `_BACKEND.simulate(net)` is the appropriate place for arguments that determine the types of data that are output/saved in a simulation. Any specific arguments against this?

Comment (ntolley) [reply hnn_core/dipole.py:66]: Fair point, my only hesitancy is tacking on a bunch of arguments to the critical functions. More of a clean/readability concern, but these seem sufficiently important.

Comment (ntolley) [reply hnn_core/dipole.py]: Decided to go with making them attributes of `Network` since they can be updated once and readout by `network_builder.py`. Realized that passing them as arguments would require a lot of copy-pasted code as I tried it.

**Comment by ntolley** [comment]
Inspired by a talk from Eve Marder, she showed off a really nice visualization of ionic current flow for single cells:
image

Thinking of adding something similar to `viz.py`. I think this is particularly relevant to characterizing degenerate parameter sets that produce the same current dipole.

**Comment by jasmainak** [comment]
go for it!

**Comment by ntolley** [comment]
Quick update: I've finished merging the recording functions. For `cell.py` I kept `insert_dipole()` separate from `record_soma()` as the former is significantly more complex. The somatic voltage and current recordings are handled by `record_soma()`. 

However in `network_builder.py` I've combined the dipole, current, and voltage aggregation into a single `aggregate_data()` function as they all require looping over every gid.

File: hnn_core/cell.py
Comment (jasmainak) [inline hnn_core/cell.py:250]: would something bad happen if you remove the `try`-`except` ? If not, I'd get rid of it

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:32]:
```suggestion
   cell_response : CellResponse
```

?

File: hnn_core/network_builder.py
Comment (jasmainak) [inline hnn_core/network_builder.py:635]: dict_currents -> rec_i ?

File: hnn_core/cell.py
Comment (ntolley) [reply hnn_core/cell.py:250]: I don't think so... this is code left over from the original current recording setup. If nothing changes I'll convert it to mimic the voltage recording naming schemes (including the variable name changes below).

File: hnn_core/network_builder.py
Comment (ntolley) [reply hnn_core/network_builder.py:635]: May as well. It is a dictionary of `h.Vector()` recordings unlike `vsoma` but anyone who is digging around this code can quickly figure that out.

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:32]: I'll try it out. This rename will change quite a lot of the code, but we may as well keep the naming consistent until we think of a better alternative.

**Comment by ntolley** [comment]
@jasmainak @rythorpe I've pretty satisfied with the changes so far, definitely feel free to suggest modifications. Once this is merged I'll start a PR converting the `CellResponse` attribute into a list stored by `Network`.

Also some background for voltage/spike timing test in `test_dipole.py`. It's essentially seeing if the spike times fall in the range of increased somatic voltage for a particular gid.

File: examples/plot_simulate_evoked.py
Comment (jasmainak) [inline examples/plot_simulate_evoked.py:63]: humm ... now it seems this should be `plot_spikes`. We need to make all this consistent, but I'm okay if you want to push it to another PR

File: hnn_core/tests/test_dipole.py
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:67]: does this work for both the backends? Can you add a test to check that?

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:60]: I'm a bit suspicious if this will work for MPI backend.

Comment (ntolley) [reply hnn_core/dipole.py:60]: Really good catch, my local machine was defaulting to the JobLib backend since I didn't specify `n_procs`. MPI produces returns data identical to `record_vsoma=False`. 

Not sure I totally understand the behavior but will dig in tomorrow. Do you have a grasp of the issue? I'm guessing `net._record_vsoma` is not persisting after being passed to the backend.

Comment (jasmainak) [reply hnn_core/dipole.py:60]: my hunch is that the `MPIBackend` only passes the `params` dictionary not the whole object. I think some effort spent in cleaning up the backends and codepaths would hold us in good stead in the long run.

Comment (ntolley) [reply hnn_core/dipole.py:60]: Think I found what you're talking. The params are pickled in `parallel_backends.py` [here](https://github.com/jonescompneurolab/hnn-core/blob/b50f91dcf7c0cdeae87189c9a1dc97f56bddc0b4/hnn_core/parallel_backends.py#L356) and `mpi_child.py` instantiates a new network based solely on the `params` [here](https://github.com/jonescompneurolab/hnn-core/blob/b50f91dcf7c0cdeae87189c9a1dc97f56bddc0b4/hnn_core/mpi_child.py#L103). Short term solution would be reverting back to tacking `record_vsoma` and `record_isoma` onto the `params` dictionary. 

Can you open an issue/PR for a discussion on what needs to be changed for the backends? I don't feel I have a good sense of what's good or bad about the current implementation.

Comment (jasmainak) [reply hnn_core/dipole.py:60]: Take a look at #142 but you need to muck with the code to realize the limitations/issues. We can't depend on one person for that code. For example, joblib backend allows any picklable object, not just `params` (will open an issue and tag you). Try passing the `record_vsoma` to `_Backend.simulate()` and you'll see how nested and deep the calls quickly get. Shunting through `net.params` is [code smell](https://en.wikipedia.org/wiki/Code_smell)

Comment (ntolley) [reply hnn_core/dipole.py:60]: Do you want to pause merging this PR until the parallel issues are addressed? I agree it's not the best solution but I don't see clear workaround besides hardcoding the parameter passing.

Comment (jasmainak) [reply hnn_core/dipole.py:60]: no, let's move on. Are you able to make it work for both backends though?

Comment (ntolley) [reply hnn_core/dipole.py]: Yes they both work locally after reverting back to passing through `params`. I've added explicit tests for both backends in `test_dipole.py` (L68). They're a bit over complete as both involve checking the size of the recorded output, let me know if you think it should be cut down.

File: examples/plot_simulate_evoked.py
Comment (ntolley) [reply examples/plot_simulate_evoked.py]: Renamed it to `plot_spikes_raster()`. Additionally changed `plot_hist()` to `plot_spikes_hist()` for consistency.

File: examples/plot_firing_pattern.py
Comment (jasmainak) [inline examples/plot_firing_pattern.py:41]: can you say a bit about the `CellResponse` object here? What does it contain?

File: hnn_core/dipole.py
Comment (jasmainak) [reply hnn_core/dipole.py:66]: do you mind leaving a comment where these parameters are used again? I guess in `cell.py` or `_simulate_single_trial`?

File: hnn_core/tests/test_dipole.py
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:70]: define `n_trials = 2` above and compare to `n_trials`

File: hnn_core/tests/test_dipole.py
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:74]: what's 8?

**Review by jasmainak — APPROVED** [review]
PR looks okay for now, let's iterate more in future PRs. @rythorpe do you want to take a look and merge if you're happy?

don't forget to update `whats_new.rst` ?

**Comment by jasmainak** [comment]
you might also need to update `api.rst`

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:34]: ?? something is missing ...

File: hnn_core/tests/test_dipole.py
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:87]: add comment above L87 what this test is doing in plain English

File: hnn_core/tests/test_dipole.py
Comment (ntolley) [reply hnn_core/tests/test_dipole.py:74]: It's just the gid `vsoma` is recorded from. I'll assign explicit variables along with `n_trials` above.

**Comment by ntolley** [comment]
Glad I caught the MPI aggregation bug before merging... Previously the MPI backend would only store cell recordings from node 0. To avoid similar problems in the future I've added a test to explicitly compare the outputs of the MPI and JobLib backends.

@jasmainak what do you have in mind for `api.rst`?

**Comment by jasmainak** [comment]
> @jasmainak what do you have in mind for api.rst?

don't you need to rename `Spikes` to `CellResponse` there?

**Comment by jasmainak** [comment]
Other than `api.rst`, PR LGTM, I let @rythorpe merge if he's happy

File: hnn_core/network.py
Comment (rythorpe) [inline hnn_core/network.py:403]: Aren't these actually lists (n_trials,) of dicts?

File: hnn_core/network.py
Comment (rythorpe) [inline hnn_core/network.py:404]: In keeping with our previous convention to use built-in python objects when possible, could we change `times` to type list?

Comment (rythorpe) [reply hnn_core/network.py:404]: This is super nitpicky so we can also save this for the next PR if we want.

**Review by rythorpe — APPROVED** [review]
Aside from my 2 very minor comments re: the `CellResponse` docstring, this looks really good @ntolley! As soon as you respond, let's merge!

**Comment by rythorpe** [comment]
Oh, and also update `api.rst`!

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:404]: Eh this is tough. There is a lot of code the presumes `times` is an ndarray for logical indexing and other conveniences. There is even an explicit test to ensure that `times` is instantiated as an array and not a list in [`test_network.py`](https://github.com/jonescompneurolab/hnn-core/blob/b50f91dcf7c0cdeae87189c9a1dc97f56bddc0b4/hnn_core/tests/test_network.py#L49). 

The change would just involved recasting with `np.array()` in a variety of placed around the code. I'm fine with changing to a list in this PR but I wonder if `times` is worth making an exception...

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:403]: Good catch! Updated.

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:404]: I believe `times` is an array in MNE. It is indeed quite convenient in many places.

Comment (rythorpe) [reply hnn_core/network.py:407]: Let's leave it then.

**Comment by rythorpe** [comment]
Thanks @ntolley!

**Comment by jasmainak** [comment]
Great stuff! Next PR is to add the somatic current visualization to an example?

**Comment by ntolley** [comment]
I was planning to leave that until after converting `CellResponse` to a list. Do you think that should wait until the remaining integration PR's are merged?

**Comment by jasmainak** [comment]
I'm guessing the `CellResponse` PR will take a while to merge? So it should be fine to get started on that. Meanwhile we should try to get @kohl-carmen 's PR merged. @kohl-carmen has unwittingly stepped on the `MPIBackend` landmine ... hope to get it resolved soon.

**Pull request merged**
